src/pages/index.js

```diff
@@ -63,27 +100,102 @@ function handleImageClick(cardLink, cardCaption) {
   imagePopup.open(cardLink, cardCaption);
 }
 
+function handleAvatarClick() {
+  editAvatarPopup.open();
+}
+
+function handleDeletePopup(cardId, evt) {
+  deleteVerifyPopup.open();
+  cardIdForDeletion = cardId;
+  cardToDelete = evt.target.closest(".card");
+}
+
+function removeLikeFromServer(cardId, likeButton) {
+  api
+    .removeLikeFromCard(cardId)
+    .then(likeButton.classList.remove("card__liked_active"))
+    .catch((err) => console.error(err));
+}
+
+function likeToServer(cardId, likeButton) {
+  api
+    .addLikeToCard(cardId)
+    .then(likeButton.classList.add("card__liked_active"))
+    .catch((err) => console.error(err));
+}
+
 function createCard(imageInfo) {
-  const newCard = new Card(imageInfo, "#card-template", handleImageClick);
+  const newCard = new Card(
+    imageInfo,
+    "#card-template",
+    handleImageClick,
+    handleDeletePopup,
+    removeLikeFromServer,
+    likeToServer
+  );
   const cardElement = newCard.generateCard();
   return cardElement;
 }
 
+function handleSubmit(request, popupInstance, loadingText = "Saving...") {
+  popupInstance.renderLoading(true, loadingText);
+  request()
+    .then(() => {
+      popupInstance.close();
+    })
+    .catch(console.error)
+    .finally(() => {
+      popupInstance.renderLoading(false);
+    });
+}
+
 function handleAddImage(imageInfo) {
-  const addedCardElement = createCard(imageInfo);
-  cardSection.addItem(addedCardElement, "prepend");
-  addFormPopup.close();
+  function makeRequest() {
+    return api.addCard(imageInfo).then((res) => {
+      const addedCardElement = createCard(res);
+      cardSection.addItem(addedCardElement, "prepend");
+    });
+  }
+  handleSubmit(makeRequest, addFormPopup);
   addFormPopup.clearForm();
   formValidators["add"].disableSubmitButton();
```

Comment:
These 2 lines should be in `then` in line 156


You need to **reset** the form only  in block `then` not to do that immediately after `submit` because when there is an error the popup will still be opened and the user can try to send the data again to the server without having to type the data into the inputs once more